src/mqtt/client.rs

```diff
@@ -911,7 +911,53 @@ impl asynch::Connection for EspAsyncMqttConnection {
     }
 }
 
-static ERROR: EspError = EspError::from_infallible::<ESP_FAIL>();
+
+#[derive(Debug)]
+pub enum MqttClientError {
+    None,
+    TcpTransport {
+        esp_tls_last_esp_err: i32,
+        esp_tls_stack_err: i32,
+        esp_tls_cert_verify_flags: i32,
+    },
+    ConnectionRefused {
+        code: u32
+    },
+    SubscribeFailed,
+    UnknownError { code: u32 }
+}
+
+
+#[repr(transparent)]
+pub struct EspMqttEventErrorWrapper(esp_mqtt_error_codes_t);
+
+
+impl EspMqttEventErrorWrapper {
+    pub fn error_type(&self) -> esp_mqtt_error_type_t {
+        self.0.error_type
+    }
+
+    pub fn resolve_error(&self) -> MqttClientError {
+        #[allow(non_upper_case_globals)]
+        match self.error_type() {
+            esp_mqtt_error_type_t_MQTT_ERROR_TYPE_NONE => MqttClientError::None,
+            esp_mqtt_error_type_t_MQTT_ERROR_TYPE_TCP_TRANSPORT => MqttClientError::TcpTransport { 
+                esp_tls_last_esp_err: self.0.esp_tls_last_esp_err, 
+                esp_tls_stack_err: self.0.esp_tls_stack_err, 
+                esp_tls_cert_verify_flags: self.0.esp_tls_cert_verify_flags, 
+            },
+            esp_mqtt_error_type_t_MQTT_ERROR_TYPE_CONNECTION_REFUSED => MqttClientError::ConnectionRefused { code: self.0.connect_return_code },
+            esp_mqtt_error_type_t_MQTT_ERROR_TYPE_SUBSCRIBE_FAILED => MqttClientError::SubscribeFailed,
+            _ => unreachable!(),
```

Comment:
It might be unreachable now but if they ever introduce any new type in the future we would have to remember to adjust it.

To be on the save side and not panic in this case (or e.g. in the case they forgot to document a variant), i would return the `MqttClientError::UnknownError` variant in this case.